File: leetcode/no0377_Combination_Sum_IV.go
package main

import "fmt"

func combinationSum4(nums []int, target int) int {

	pathsToTargetSum := make([]int, target+1)
	pathsToTargetSum[0] = 1

	for value := 1; value <= target; value++ {
		for _, num := range nums {
			diff := value - num
			if diff >= 0 {
				pathsToTargetSum[value] += pathsToTargetSum[diff]
			}
		}
	}
	return pathsToTargetSum[target]
}

func main() {
	input := []int{1, 2, 3}
	target := 4
	result := combinationSum4(input, target)
	fmt.Println("OUTPUT:", result)
}

File: leetcode/no0377_Combination_Sum_IV.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var combinationSum4 = function (nums, target) {

    console.log("INPUT:", nums, target);
    let dp = []; // dp[i] # of combinations sum up to i
    dp[0] = 1;

    console.log("Before Start:", dp);

    for (let value = 1; value <= target; ++value) {

        dp[value] ? null : dp[value] = 0;

        console.log("check value:", value);

        nums.forEach((choosedNum) => {

            console.log("Take", choosedNum, "from", nums);
            let remaining = value - choosedNum;
            console.log(`value: ${value} - num: ${choosedNum} = ${remaining}`);
            if (remaining >= 0) {
                dp[value] += dp[remaining];
            }
            console.log(dp)
        })
    }
    let result = dp[target];
    console.log(dp);
    console.log("OUTPUT:", result);
    return result;
};

combinationSum4([1, 2, 3], 4)